flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/internal/converter/AbstractJdbcRowConverter.java

```diff
@@ -18,54 +18,255 @@
+		final int sqlType = JdbcTypeUtil.typeInformationToSqlType(TypeConversions.fromDataTypeToLegacyInfo(
+			TypeConversions.fromLogicalToDataType(type)));
+		return (val, index, statement)  -> {
+			if (val == null || val.isNullAt(index) || LogicalTypeRoot.NULL.equals(type.getTypeRoot())) {
+				statement.setNull(index + 1, sqlType);
+			} else {
+				jdbcSerializationConverter.serialize(val, index, statement);
+			}
+		};
+	}
+
+	protected JdbcSerializationConverter createExternalConverter(LogicalType type) {
+		switch (type.getTypeRoot()) {
+			case BOOLEAN:
+				return (val, index, statement) -> statement.setBoolean(index + 1, val.getBoolean(index));
+			case TINYINT:
+				return (val, index, statement) -> statement.setByte(index + 1, val.getByte(index));
+			case SMALLINT:
+				return (val, index, statement) -> statement.setShort(index + 1, val.getShort(index));
+			case INTEGER:
+			case INTERVAL_YEAR_MONTH:
+				return (val, index, statement) -> statement.setInt(index + 1, val.getInt(index));
+			case BIGINT:
+			case INTERVAL_DAY_TIME:
+				return (val, index, statement) -> statement.setLong(index + 1, val.getLong(index));
+			case FLOAT:
+				return (val, index, statement) -> statement.setFloat(index + 1, val.getFloat(index));
+			case DOUBLE:
+				return (val, index, statement) -> statement.setDouble(index + 1, val.getDouble(index));
+			case CHAR:
+			case VARCHAR:
+				// value is BinaryString
+				return (val, index, statement) -> statement.setString(index + 1, val.getString(index).toString());
+			case BINARY:
+			case VARBINARY:
+				return (val, index, statement) -> statement.setBytes(index + 1, val.getBinary(index));
+			case DATE:
+				return (val, index, statement) ->
+					statement.setDate(index + 1, Date.valueOf(LocalDate.ofEpochDay(val.getInt(index))));
+			case TIME_WITHOUT_TIME_ZONE:
+				return (val, index, statement) ->
+					statement.setTime(index + 1, Time.valueOf(LocalTime.ofSecondOfDay(val.getInt(index) / 1000L)));
+			case TIMESTAMP_WITH_TIME_ZONE:
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+				final int timestampPrecision = ((TimestampType) type).getPrecision();
+				return (val, index, statement) ->
+					statement.setTimestamp(index + 1, val
+						.getTimestamp(index, timestampPrecision)
+						.toTimestamp());
+			case DECIMAL:
+				final int decimalPrecision = ((DecimalType) type).getPrecision();
+				final int decimalScale = ((DecimalType) type).getScale();
+				return (val, index, statement) ->
+					statement.setBigDecimal(index + 1, val
+						.getDecimal(index, decimalPrecision, decimalScale)
+						.toBigDecimal());
+			case ARRAY:
+				//note: dialect need implements the conversion from ArrayData to JDBC Array if the dialect supports array.
+				return (val, index, statement) -> {
+					throw new IllegalStateException(
+						String.format("JDBC:%s do not support write ARRAY type.", converterName()));
+				};
```

Comment:
If the dialect doesn't support ARRAY, please throw the exception at compile stage, rather than at runtime.

Reply:
From now,  we use same RowType to build a JdbcRowConverter that contains serialize and deserialize converter, but Postgres dialect support PgArray in source but not support in sink(that's the reason), if we throw the exception at compile stage, the deserialize converter can not use too.